doc/developer/design/20231213_compute_operator_hydration_status_tracking.md

```diff
@@ -0,0 +1,192 @@
+To reduce complexity in the Compute code, we choose not to attempt this approach.
+
+Instead, we propose adding a hydration status logging operator at the end of each LIR region.
+These operators will observe frontier progress in their inputs and report hydration events via loggers connected to the `ComputeState`.
+There is already precedence for such operators, namely the ones that log import frontiers and arrangement sizes.
+
+### Reporting Introspection Data
+
+Existing replica introspection data is exposed through per-replica logging dataflows that export introspection arrangements.
+There are two issues with this approach:
+
+1. In multi-replica clusters, different replicas usually have different data in these arrangements.
+   As a result, introspection arrangements can only be queried by special replica-targeted queries.
+   To collect introspection data from all replicas, clients have to enumerate the existing replicas and issue a separate query for each, yielding bad UX.
+2. When a replica is slow to respond to queries, as if often the case during hydration, it will also be slow to respond to introspection queries.
+   This is particularly annoying when one wants to use introspection queries to monitor hydration progress.
+
+To avoid these issues, we propose reporting operator hydration information to the compute controller through `ComputeResponse`s instead.
+The compute controller can then write this information into storage-managed collections and expose it to SQL clients that way.
+This solves issue (1) as introspection information from all replicas ends up in the same collection and can be retrieved with a single query.
+Issue (2) is mostly solved because the time logging events spend on the replica is reduced.
+It is still the case that when a replica is completely unresponsive, it likely won't schedule the hydration logging operators, so no new introspection data will make it to the compute controller either.
+However, even in these (hopefully rare) cases introspection data previously written can still be queried.
+
+### SQL-level Changes
+
+This design proposes adding two new relations in which operator hydration status information is exposed:
+
+```
+mz_internal.mz_compute_operator_hydration_statuses_per_worker {
+    object_id: text,
+    physical_plan_id: uint8,
+    replica_id: text,
+    worker_id: uint8,
+    hydrated: bool,
+}
+
+mz_internal.mz_compute_operator_hydration_statuses {
+    object_id: text,
+    physical_plan_id: uint8,
+    replica_id: text,
+    hydrated: bool,
+}
+```
+
+... where the former is a storage-managed collection (aka. `BuiltinSource`) while the latter is an aggregating view over the former.
+
+Furthermore, we propose adding a new `EXPLAIN` output modifier, `plan_ids`, that toggles rendering of LIDs in physical plan outputs.
```

Comment:
nit: I have a slight preference for this modifier to be called `node_id` or `operator_id` instead of `plan_id`. As I said, even though `EXPLAIN` currently does not expose that information, we have statement types that produce a description of more then one `objects_to_build`, and I would probably use `plan_id` to name an identifier for each plan in the `objects_to_build` entry.

Reply:
Fine with me! I chose `plan_id` to be consistent with `physical_plan_id` in the SQL relations, but we can rename that to `physical_plan_node_id`. Speaking about "nodes" is probably more intuitive for users.

I'd just like to avoid calling these "operator IDs" because we already have Timely operator IDs and these are not the same.

Reply:
:+1: for `plan_node_id` in catalog relations and `node_id` for code variables and `EXPLAIN` output annotations.